
{% macro sample_qc(params) -%}
    <h2>Sample QC</h2>
    <p  class="font5">Genomic DNA (gDNA) samples are evaluated for quantity using Quant-iT Picogren reagent, Lambda Standard DNA
    and a Molecular Devices, Spectramax XPS Gemini plate reader.  The quality of the gDNA samples are evaluated
    using an AATI Fragment Analyzer and the Standard Sensitivity Genomic DNA Analysis Kit. Genomic DNA samples
    found to have a total amount of >1000ng and a quality score >5 pass sample QC. Based on the quantification results,
    gDNA samples are pre-normalised to fall within the 5-40ng/uL concentration range required for Illumina SeqLab
    TruSeq Nano library preparation method using the Hamilton MicroLab STAR.</p>
{%- endmacro %}


{% macro library_prep(params) -%}
    <h2>Library Preparation</h2>
    <p class="font5">Next Generation sequencing libraries are prepared using an Illumina SeqLab specific TruSeq Nano High
    Throughput library preparation kit in conjunction with the Hamilton MicroLab STAR and Clarity LIMS X (4.2?)
    Edition. The 200ng gDNA sample input is normalised to the concentration and volume required for the Illumina
    TruSeq Nano library preparation kit, then sheared to a 450bp mean insert size using a Covaris LE220
    focused-ultrasonicator. The inserts are blunt ended, A-tailed, size selected, TruSeq adapters are ligated
    onto the ends of each fragment before being PCR amplified.</p>
{%- endmacro %}


{% macro library_qc(params) -%}
    <h2>Library QC</h2>
    <p class="font5">The insert size for each library is evaluated using the Caliper GX Touch with a HT DNA 1k/12K/HI SENS LabChip
    and HT DNA HI SENS Reagent Kit to ensure that the mean fragment sizes fall between 530bp and 730bp.
    The concentration of each library is calculated using a Roche LightCycler 480 and a Kapa Illumina Library
    Quantification kit and Standards to ensure that each libraries concentration is between 5.5nM and 40nM.</p>
{%- endmacro %}


{% macro sequencing(params) -%}
    <h2>Sequencing</h2>
    <p class="font5">The libraries are normalised to 1.5nM and are denatured for clustering and sequencing at 300pM using a
    Hamilton MicroLab STAR with Genologics Clarity LIMS X (4.2?) Edition. Libraries are clustered onto a
    HiSeqX Flow cell v2.5 on cBot2s and the clustered flow cell is transferred to a HiSeqX for sequencing
    using a HiSeqX Ten Reagent kit v2.5.</p>
{%- endmacro %}

{% macro non_human_bioinf(params) -%}
    <h2>Bioinformatics analysis</h2>
    <p class="font5">Demultiplexing is performed using bcl2fastq ({{ params['bcl2fastq_version'] }}), allowing 1 mismatch when assigning reads
    to barcodes. Adapters (Read1: {{ params['adapter1'] }}, Read2: {{ params['adapter2'] }}) are trimmed during the demultiplexing
    process.</p>
{%- endmacro %}

{% macro bos_taurus_bioinf(params) -%}
    <p class="font5">bwa mem ({{ params['bwa_version'] }}) is used to align the raw reads to the {{ params['genome_version'] }} genome, and the Genome Analysis ToolKit ({{ params['gatk_version'] }}) for the indel realignment
    and base recalibration. The genotype likelihoods are calculated using Genome Analysis Toolkit
    ({{ params['gatk_version'] }}) HaplotypeCaller creating a final gvcf file.</p>
{%- endmacro %}

{% macro human_bioinf(params) -%}
    <h2>Bioinformatics analysis</h2>
    <p class="font5">Demultiplexing is performed using bcl2fastq ({{ params['bcl2fastq_version'] }}), allowing 1 mismatch when assigning reads
    to barcodes. Adapters (Read1: {{ params['adapter1'] }}, Read2: {{ params['adapter2'] }}) are trimmed during the demultiplexing
    process. BCBio-Nextgen ({{ params['bcbio_version'] }}) is used to perform alignment, bam file preparation and variant
    detection. Variant data is used to verify sample identity by comparing to SNPs genotyped using the
    Life Technologies QuantStudio 12K, 32A TaqMan SNP Openarrays. BCBio uses bwa mem ({{ params['bwa_version'] }}) to align the raw reads to the {{genome_name}}
    ({{ params['genome_version'] }}) genome, then samblaster ({{ params['samblaster_version'] }}) to mark the duplicated fragments, and
    the Genome Analysis ToolKit ({{ params['gatk_version'] }}) for the indel realignment and base recalibration. The genotype
    likelihoods are calculated using Genome Analysis Toolkit ({{ params['gatk_version'] }}) HaplotypeCaller creating a final
    gvcf file.</p>
{%- endmacro %}


{% macro yield_chart(params) -%}
    <h2>Yield</h2>
    <img id="chart1" class="charts" src="{{ params['yield_chart'] }}" />
    <p class="font5">Yield is calculated as the yield in gigabases (Gb) after adapters have been
    trimmed during processing. The yield Q30 is the quantity of reads
     which have at least 75% of bases with a quality score of 30 or higher.</p>
{%- endmacro %}

{% macro mapping_duplicates_chart(params) -%}
    <h2>Read Mapping and Duplicates</h2>
    <img id="chart2" class="charts" src="{{ params['mapping_duplicates_chart'] }}" />
    <p class="font5">Optical and PCR duplicates can be generated during library preparation, clustering, or be due to
    sample quality. The percentage of duplicates is calculated by comparing the number of duplicate reads
    to the total read count. Duplicates are identified during read mapping to prevent the overestimation
    of coverage, therefore, the percentage of paired reads aligned to the reference genome includes
    non-duplicate reads only. A duplicate rate of < 30% is required for each sample, except in circumstances
    where a very high yield achieves sufficient coverage of the genome despite a high duplicate rate.</p>
{%- endmacro %}

{% macro fastq() -%}
    <h2>Fastq</h2>
    <p class="font5">Files ending in 'fastq.gz' are in gzipped Sanger fastq format, and contain the raw sequence data after
    demultiplexing. Each sample has two files: an 'R1.fastq.gz' and an 'R2.fastq.gz', respectively containing
    the first and second reads of each read pair.</p>
    <p class="font5"><a href="https://en.wikipedia.org/wiki/FASTQ_format">Fastq specification</a></p>
{%- endmacro %}

{% macro bam() -%}
    <h2>BAM</h2>
    <p class="font5">Files ending in '.bam' are in BAM format, and contain the reads aligned to the genome and ready for
    variant calling. Each sample has one BAM file plus a '.bam.bai' file, which is the BAM index. This
    allows some programmes fast random access to any part of the BAM file.</p>
    <p class="font5"><a href="https://samtools.github.io/hts-specs/SAMv1.pdf">Sam/Bam specification</a></p>
{%- endmacro %}

{% macro vcf() -%}
    <h2>VCF</h2>
    <p class="font5">Files ending in '.vcf.gz' are gzipped VCF files, and contain variants dectected between the sample and
    the reference genome. There is also a 'g.vcf.gz' file, which contains genotype likelihoods across the
    genome regardless of the presence of a variant. Files ending in 'vcf.gz.tbi' are Tabix indexes
    allowing fast random access to any part of their corresponding VCF.</p>
    <p class="font5"><a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF specification</a></p>
{%- endmacro %}
